cupy/cuda/cupy_thrust.cu

```diff
@@ -1,476 +1,62 @@
-                               dp_data_last,
-                               dp_idx_first);
-        } else {
-            // Generate key indices.
-            dp_keys_first = thrust::device_pointer_cast(static_cast<size_t*>(keys_start));
-            dp_keys_last  = thrust::device_pointer_cast(static_cast<size_t*>(keys_start) + size);
-            transform(cuda::par(alloc).on(stream_),
-                      thrust::make_counting_iterator<size_t>(0),
-                      thrust::make_counting_iterator<size_t>(size),
-                      thrust::make_constant_iterator<ptrdiff_t>(shape[ndim-1]),
-                      dp_keys_first,
-                      thrust::divides<size_t>());
-
-            stable_sort_by_key(
-                cuda::par(alloc).on(stream_),
-                make_zip_iterator(make_tuple(dp_keys_first, dp_data_first)),
-                make_zip_iterator(make_tuple(dp_keys_last, dp_data_last)),
-                dp_idx_first);
-        }
-    }
-};
+#include "cupy_thrust.inl"
+
+
+namespace cupy {
+  std::function<cupy::SortTargetT> sort_targets[CUPY_NUM_TYPES] = {
+    thrust_sort_CUPY_TYPE_INT8      ,
+    thrust_sort_CUPY_TYPE_UINT8     ,
+    thrust_sort_CUPY_TYPE_INT16     ,
+    thrust_sort_CUPY_TYPE_UINT16    ,
+    thrust_sort_CUPY_TYPE_INT32     ,
+    thrust_sort_CUPY_TYPE_UINT32    ,
+    thrust_sort_CUPY_TYPE_INT64     ,
+    thrust_sort_CUPY_TYPE_UINT64    ,
+    thrust_sort_CUPY_TYPE_FLOAT16   ,
+    thrust_sort_CUPY_TYPE_FLOAT32   ,
+    thrust_sort_CUPY_TYPE_FLOAT64   ,
+    thrust_sort_CUPY_TYPE_COMPLEX64 ,
+    thrust_sort_CUPY_TYPE_COMPLEX128,
+    thrust_sort_CUPY_TYPE_BOOL      ,
+  };
+
+  std::function<cupy::LexSortTargetT> lexsort_targets[CUPY_NUM_TYPES] = {
+    thrust_lexsort_CUPY_TYPE_INT8      ,
+    thrust_lexsort_CUPY_TYPE_UINT8     ,
+    thrust_lexsort_CUPY_TYPE_INT16     ,
+    thrust_lexsort_CUPY_TYPE_UINT16    ,
+    thrust_lexsort_CUPY_TYPE_INT32     ,
+    thrust_lexsort_CUPY_TYPE_UINT32    ,
+    thrust_lexsort_CUPY_TYPE_INT64     ,
+    thrust_lexsort_CUPY_TYPE_UINT64    ,
+    thrust_lexsort_CUPY_TYPE_FLOAT16   ,
+    thrust_lexsort_CUPY_TYPE_FLOAT32   ,
+    thrust_lexsort_CUPY_TYPE_FLOAT64   ,
+    thrust_lexsort_CUPY_TYPE_COMPLEX64 ,
+    thrust_lexsort_CUPY_TYPE_COMPLEX128,
+    thrust_lexsort_CUPY_TYPE_BOOL      ,
+  };
+
+  std::function<cupy::ArgSortTargetT> argsort_targets[CUPY_NUM_TYPES] = {
+    thrust_argsort_CUPY_TYPE_INT8      ,
+    thrust_argsort_CUPY_TYPE_UINT8     ,
+    thrust_argsort_CUPY_TYPE_INT16     ,
+    thrust_argsort_CUPY_TYPE_UINT16    ,
+    thrust_argsort_CUPY_TYPE_INT32     ,
+    thrust_argsort_CUPY_TYPE_UINT32    ,
+    thrust_argsort_CUPY_TYPE_INT64     ,
+    thrust_argsort_CUPY_TYPE_UINT64    ,
+    thrust_argsort_CUPY_TYPE_FLOAT16   ,
+    thrust_argsort_CUPY_TYPE_FLOAT32   ,
+    thrust_argsort_CUPY_TYPE_FLOAT64   ,
+    thrust_argsort_CUPY_TYPE_COMPLEX64 ,
+    thrust_argsort_CUPY_TYPE_COMPLEX128,
+    thrust_argsort_CUPY_TYPE_BOOL      ,
```

Comment:
@emcastillo Here we collect an array of func ptr for runtime dispatching. I feel this is the only way, but any chance you can think of some template tricks that can achieve this (but without instantiating the template implicitly)?